### ISSUE TITLE
Consider exposing access to 'name' table on hb_face_t or hb_font_t

### ISSUE THREAD

**Issue by drott** [behdad/harfbuzz#432]
For fixing the `@font-face { src: local(xyz) }` matching, we need to look at postscript name or full name in the `name` table. It seems like `hb-ot-name.hh` already has some parsing for this. 

If accessing such information would be available on `hb_face_t`, that would be easiest for reusing in Blink and allow us to at least validate a match for `local()`.

**Comment by roozbehp** [comment]
Useful in Android too, although for a different reason (debugging).

**Comment by KrasnayaPloshchad** [comment]
What happened if this table works with 'stat' table?

**Comment by behdad** [comment]
My current thinking on the API for this:

- We want the actual name fetcher to have two variants, one for UTF-8, another for UTF-16.  The cleanest, though not most user-friendly, would be the fetcher only takes an index into the list of names. The problem is, this is NOT the nameID, but an otherwise useless index that the user first needs to fetch using following API,

- The matching API takes nameID, platformID, platEncID, languageID.  Each of platformID, platEncID, and languageID can each be "wildcard" (0xFFFF), to let HarfBuzz choose best.  Easiest would be for this API to return the index to be used by the fetching API, to avoid having to expose separate UTF-8 / UTF-16 versions,

- An enumeration API. My currently thinking is that this one simply takes start_offset and count and returns the nameID/platformID/platEncID/languageID of the name table entry at those indices.  Ie. this is very raw.

What this API does not make easy is to list all languages of a nameID. Maybe the enumeration API can have another counterpart that takes a nameID?

cc @khaledhosny 

- API to fetch name by an index into the list of names.

**Comment by behdad** [comment]
I thought more about this. I'm leaning towards NOT exposing the platform/encoding/language madness, and provide a higher-level API that works with hb_language_t only.

**Comment by twardoch** [comment]
Just 4 remarks (more observations then recommendations, as I don't have opinions on the API but perhaps it'll be useful): 

- I agree that smart stuff like falling back from name ID 17 to 2 may be out of scope, so reading each name ID should be treated the same

- What the API could do is to simplify the language matching. The 3.1 name branch uses the old Win 32 API language numbers, the 1.0 or 0.3 branches use something else; it would be nice if the API tied them into HarfBuzz language tags

- In reality, the "English" names are often treated as default, so something like "undefined language" or "generic language" should fetch the English names

- The 1.x name IDs aren't Unicode, they're encoded in a mixture of old encodings. We tackled this in fontTools 3 years ago* — not sure how you want to deal with it here

https://github.com/fonttools/fonttools/issues/236

**Comment by khaledhosny** [comment]
What  Behdad said.

What I did before (in Scribus, but that is a horrible code base so I’m not providing links), is prefer platform 3 (Windows) > 0 (Unicode) > 1 (Macintosh), initially I preferred Unicode first but many fonts had buggy names there so Windows names seem to be more tested. For Windows platforms prefer encoding 10 (Unicode full repertoire) > 1 (Unicode BMP). Anything else was picked randomly (probably by the order they are present in the font) and I didn’t care much since they all were equally bad.

**Comment by behdad** [comment]
> * The 1.x name IDs aren't Unicode, they're encoded in a mixture of old encodings. We tackled this in fontTools 3 years ago* — not sure how you want to deal with it here

Right.  I was thinking about ignoring all non-Unicode ones except for MacRoman.  Maybe just the ASCII part of MacRoman.

**Comment by twardoch** [comment]
And, when the user asks for "give me name ID 17 In "Japanese", and there is no Japanese — should the API fetch the English aka default name instead?

**Comment by behdad** [comment]
> And, when the user asks for "give me name ID 17 In "Japanese"", and there is no Japanese — should the API fetch the English aka default name instead?

I think we can leave it to client to do the fallback.  Am not sure though.

**Comment by twardoch** [comment]
I agree about ignoring noon-Roman 1.x. There are virtually no fonts that have 1.x but have no 3.1 or 0.3. But: remember that a lot of Apple system fonts bundled with macOS/iOS have 0.3 but no 3.1. I think Apple just did it to "annoy" people who'd like to move these fonts to their Windows machines or something :)

**Comment by khaledhosny** [comment]
If a client has a list of locales to use not just (I know that at least Firefox supports this) then it might want to try a different fallback language before English. Doing fallback here will make unnecessarily harder for the client to use a different fallback strategy.

**Comment by khaledhosny** [comment]
Gettext also supports a list of fallback locales and one can certainly choose a different fallback locale other than English (though English is always the last resort).

**Comment by behdad** [comment]
> There are virtually no fonts that have 1.x but have no 3.1 or 0.3.

Apple Color Emoji has MacRoman only for English.  I think they use it as "ASCII" since it saves bytes.

**Comment by behdad** [comment]
Okay.  I thought more about this.  Here's what I suggest.  It includes low-level and high-level API, so should be good for every use.  Note, fetching all languages for a nameID is useful for fontconfig.  But then again, if it's going to be expensive (ie. O(N)) for each one, then fontconfig should just use the lower-level API and sort it out itself (like it does right now).  So, here's current proposal, without the list-languages...

- get_count returns total number of name table entries (we call the index into this a name index.  Contrast with nameID, which we have a `hb_name_it_t` type for).

- get_entry returns a struct with unsigned int platformID, encodingID, languageID, hb_name_id_t nameID.

The above are the low-level API.

- get_name takes a nameID and language.  Tries finding name for requested language.  If language is INVALID, tries English.  If not found, retries with English.  Returns true if fallback was NOT used.  Maybe two versions, one utf8, another utf16.

I guess that's the base.

We can also expose a few helpers for clients who want to process low-level, but I'm not sure about these:

- is_unicode takes a name_entry_t and returns true if encoding is Unicode.  Looks like all Unicode encodings in name table are (subset of) UTF-16.

- get_encoding takes a name_entry_t, or just platformID, encodingID, languageID, and returns a encoding name.  For this to be useful, we also nee:

- get_data that would take name index and return raw bytes.

Let's agree on which parts we want so we don't waste @ebraminio's efforts ;)

**Comment by twardoch** [comment]
Ok, my expertise ends here, so good luck!

**Comment by behdad** [comment]
> Ok, my expertise ends here, so good luck!

Thanks.

Any other low-hanging (or not) HB requests you have, while we are all active?

**Comment by khaledhosny** [comment]
I’d go with the high level API until there is actual demand for the low level ones (the only two times I ever needed to get entries from the `name` table, the proposed high level API would have been all I needed, FreeType API was cumbersome to use in contrast).

**Comment by behdad** [comment]
Fair enough.  Let's just add an enum and one (two, UTF-8 and UTF-16) to fetch nameID for language...

**Comment by behdad** [comment]
First step requires mapping name-table language IDs to ours.  fcfreetype.c and FreeType headers combined have this.

**Comment by behdad** [comment]
I thought more about this.  I found a full API that's I'm satisfied with:

```
HB_EXTERN hb_bool_t
hb_ot_name_get_utf8 (hb_face_t     *face,
                     hb_name_id_t  *name_id,
                     hb_language_t  language,
                     unsigned int  *text_size /* IN/OUT */,
                     char          *text      /* OUT */);

HB_EXTERN hb_bool_t
hb_ot_name_get_utf16 (hb_face_t     *face,
                      hb_name_id_t  *name_id,
                      hb_language_t  language,
                      unsigned int  *text_size /* IN/OUT */,
                      uint16_t      *text      /* OUT */);

HB_EXTERN hb_bool_t
hb_ot_name_get_utf32 (hb_face_t     *face,
                      hb_name_id_t  *name_id,
                      hb_language_t  language,
                      unsigned int  *text_size /* IN/OUT */,
                      uint32_t      *text      /* OUT */);


typedef struct hb_ot_name_entry_t
{
  hb_name_id_t  name_id;
  hb_language_t language;
} hb_ot_name_entry_t;

HB_EXTERN unsigned int
hb_ot_name_get_names (hb_face_t           *face,
                      hb_ot_name_entry_t **entries /* OUT */);
```

The first three simply fetch a name for a language (with fallback to English / use English is language is INVALID; returns false if fallback was used).

The last one, and that's how the fetchers work internally as well, returns a static sorted list of all available (ie. Unicode-only, preferred presentations only) names sorted by name-ID and language.  This satisfies all needs from such an API (including fontconfig's) without exposing any internal details.

Internally, upon first request to any of the APIs, we generated the sorted list and store it in name table's `accelerator_t`, and only use bsearch on that if needed.

**Comment by behdad** [comment]
(FWIW this is similar to how we handle post table internally.  Since the array there is not bsearch'able, we load and make a sorted one internally.)

**Comment by ebraminio** [comment]
Regarding hb_ot_name_get_*, what should happen if string doesn't fit the buffer, we are not returning actually length so client may never now what happen. Maybe returning the actual length on text_size when text is null? If that is what the way we like to shape our APIs we probably should revisit the not-published-yet hb_ot_layout_feature_get_characters for more consistency.

And about, hb_ot_name_get_names, should it be copied for client or exposing internal memory? If it should be copied more things needs to passed in/out I guess.

**Comment by ebraminio** [comment]
Can't we agree on exposing only utf8 version? Now that we should do some conversion anyway I guess one is enough and ASCII is really what most fonts have I guess.

**Comment by brawer** [comment]
Are you sure about `hb_language_t`? With an `ltag` table (and also `name` format 1), a font can supply names in arbitrary languages (embedded as arbitrary IETF BCP 47 tags), so a fixed enum won’t work. Admittedly this is a rather theoretical concern — in practice, most fonts contain just an US English name. Still, it might be wise to return language identifiers as IETF BCP 47 strings, not as an enum.

**Comment by jfkthame** [comment]
> ```
> HB_EXTERN hb_bool_t
> hb_ot_name_get_utf8 (hb_face_t     *face,
>                      hb_name_id_t  *name_id,
>                      hb_language_t  language,
>                      unsigned int  *text_size /* IN/OUT */,
>                      char          *text      /* OUT */);

No reason for `name_id` here to be a pointer, is there?

**Comment by behdad** [comment]
> No reason for `name_id` here to be a pointer, is there?

Right.  Fixed.

**Comment by behdad** [comment]
> > Can't we agree on exposing only utf8 version?

You will find that the utf8 version is the hardest. ;)  UTF-32 / UTF-16 are trivial.  But when we add writing to hb-utf.hh, then it's trivial to implement all three.

**Comment by ebraminio** [comment]
> If returned buffer is full (ie *len is equal to input *len - 1. One used for nul-termination always), then client can retry with larger buffer if desired.

Can we go better than this also? I guess this could confuse me if I was a HarfBuzz user to never go for that such a check and what if the character after the size is three bytes but only two bytes have left from input utf8 buffer?

**Comment by behdad** [comment]
> Can we go better than this also? I guess this could confuse me if I was a HarfBuzz user to never go for that such a check and what if the character after the size is three bytes but only two bytes have left from input utf8 buffer?

Yeah, it's not the nicest.  Maybe we should just malloc and return?  I'm fine with that.

**Comment by behdad** [comment]
Problem with malloc() is, we would then need to add API to free it, as we don't want to rely on free() in the API.

Back to provided buffer.

**Comment by jfkthame** [comment]
Can we always return the required size in `*text_size`? If the output value there is larger than was provided as input, then `text` was truncated to the provided length, but the actual size required is still returned. So a client can preflight by calling first with `*text_size == 0` to find out the required buffer size, then allocate it and call again with the correct size.

**Comment by behdad** [comment]
Yeah, guess that's the proper way to do it...

**Comment by behdad** [comment]
It's just more pain that we would need to accumulate the size even when not writing to buffer.

**Comment by jfkthame** [comment]
True. That'd be one reason to prefer UTF-16 APIs here... :) Also, I don't see this as perf-critical; but a client that doesn't want to pay the cost of the pre-scan could instead allocate a generous buffer up front and be almost certain of getting the complete result in a single call.

**Comment by behdad** [comment]
Yeah I'm not worried about the performance, just correctness and completeness.

**Comment by behdad** [comment]
Anyway, I'll start implementing this slowly.

**Comment by mhosken** [comment]
I agree you want 3 apis (8, 16, 32). 16 is trivial (apart from byte ordering), 32 is nearly trivial apart from SMP and 8 has some interesting aspects, especially on buffer overflow. I think returning false for fallback is OK but quirky. I think a more conventional approach would be to return false on overfull and pass in a bool for fallback or not. But I can sympathise with a desire to minimise attributes, and this does work.

**Comment by behdad** [comment]
I've changed my mind re fallback.  I'll let user deal with it.  If language is not set (ie. `HB_LANGUAGE_INVALID`), I'm thinking about setting it to default-language, instead of English.  Not sure.  Maybe I should do fallback and try default first then English though.

**Comment by ebraminio** [comment]
Maybe returning actual string length instead hb_bool_t still a possibility? I don't think normal clients won't care that much between an empty name record or non existent one and if someone really does we still have cheap `hb_ot_name_get_names`.

**Comment by behdad** [comment]
Okay, if not doing fallback, then returning full string length makes sense.

Only question left, if user passed HB_LANGUAGE_INVALID, should I interpret that as "en", or as current locale?  Leaning towards "en".

**Comment by behdad** [comment]
https://github.com/harfbuzz/harfbuzz/pull/1254 is now ready to be reviewed.
